packages/flutter/lib/src/painting/image_provider.dart

```diff
@@ -868,6 +887,104 @@ class ExactAssetImage extends AssetBundleImageProvider {
+  /// again after the next [DeferringImageProvider.deferDuration] elapses (or
+  /// at the end of current task queue if the duration is [Duration.zero]).
+  defer,
+
+  /// Indicates that the [DeferringImageProvider] should attempt to resolve the
+  /// image.
+  ///
+  /// After returning this action, the [DeferringImageProvider] will not call
+  /// its [DeferringImageProvider.getNextAction] callback anymore.
+  resolve,
+}
+
+/// An [ImageProvider] that can delay resolving an image using its wrapped
+/// [imageProvider] until an arbitrary condition is met.
+///
+/// For example, the [Image] widget uses this provider to defer loading images
+/// if it is the child of a scrollable that is rapidly scrolling. It cancels
+/// resolving the image if its element is evicted from the tree before scrolling
+/// ends, and otherwise waits for scrolling to slow down before actually
+/// resolving the image. This prevents the framework from over-utilizing
+/// hardware resources for images that will never be shown on screen.
+///
+/// If the image is already in the cache, e.g. from [precacheImage] or because
+/// another provider has already resolved it, the [getNextAction] callback will
+/// never fire and the cached image is reused.
+@immutable
+class DeferringImageProvider<T> extends ImageProvider<T> {
+
+  /// Creates a new [DeferringImageProvider].
+  ///
+  /// All parameters must not be null.
+  const DeferringImageProvider({
+    @required this.imageProvider,
+    @required this.getNextAction,
+  }) : assert(imageProvider != null),
+       assert(getNextAction != null);
+
+  /// The wrapped [ImageProvider] to delegate loading the image and creating
+  /// a key for it.
+  ///
+  /// The image provider's load method may never be called if resolution gets
+  /// cancelled.
+  final ImageProvider<T> imageProvider;
+
+  /// A callback used by [resolve] to determine whether resolution should be
+  /// cancelled, deferred, or executed.
+  ///
+  /// This callback is never called if the [ImageCache] already has the image
+  /// present, such as from [precacheImage] or another provider successfully
+  /// loading it.
+  final DeferringImageProviderAction Function() getNextAction;
+
+  @override
+  void resolveForStream(ImageConfiguration configuration, ImageStream stream) {
+    void deferredResolve(Object key) {
+      if (PaintingBinding.instance.imageCache.containsKey(key)) {
+        imageProvider.resolveForStream(configuration, stream);
+        return;
+      }
+      switch (getNextAction()) {
+        case DeferringImageProviderAction.defer:
+          // Get ourselves to the end of any frame callbacks at the beginning of
+          // the next frame, which should be after the scroll animation has
+          // updated our velocity value.
+          SchedulerBinding.instance.scheduleFrameCallback((_) {
+            scheduleMicrotask(() => deferredResolve(key));
+          });
+          return;
+        case DeferringImageProviderAction.cancel:
```

Comment:
cancel seems like it could stop an image from being loaded, but not once it starts loading? It should be possible to cancel NetworkRequests too, though I'm not sure if HttpClient exposes an API for it

Reply:
We can look at cancelling once a load starts, but this PR isn't trying to solve that one.

Part of the problem there is that it will be very hard once we actually start the machinery on the IO thread in the engine to cancel it before allocations have happened.

Network request handling could use some love but I think that's orthogonal to this issue. This is trying to stop us from even making the network request to begin with if we won't need it.

Reply:
I think it would be unfortunate if we had to come up with an entirely separate solution for network requests. I might propose a slightly more general API:

The deferring image provider accepts a controller/token instead of a callback. This controller maintains a state flag similar to what the callback is providing now.

The controller state can be modified via setters/methods like `cancel()` or `resume()`.

The cancellation could be propagated up through the network image provider, by exposing some functionality on ImageStream perhaps? Ideally this could eventually be wired up the HttpClient and others.

Reply:
My sense is that places we've introduced controllers (e.g. TextEditing) have really only caused more confusion and API challenges than they resolve.   I think it will also become very confusing if `cancel` behaves differently in different contexts - did I cancel before making the request? Did I cancel the active request? Did I cancel it early enough to have effect? Did I call cancel after the whole thing was already done?

I think for NetworkRequests made by the framework in general, we should create another InheritedWidget that would handle things like this (caching requests, cancelling requests, etc).  WDYT?

Reply:
I don't really agree but I'm not involved enough in the problem, so I'll defer to the rest of the framework team